src/python/gem5/resources/resource.py

```diff
@@ -884,6 +937,144 @@ def __init__(
         self._func = function
         self._params = parameters
 
+        self.get_all_resources(
+            resources=resources,
+            to_path=local_path,
+            resource_directory=resource_directory,
+            download_md5_mismatch=download_md5_mismatch,
+            clients=clients,
+            gem5_version=gem5_version,
+            quiet=quiet,
+        )
+
+    def get_all_resources(
+        self,
+        resources: Dict[str, Dict[str, str]],
```

Comment:
I don't understand this. You're saying when the Workload is instantiated you get all its constituent resources (downloads and all). Isn't reintroducing a problem here?  Getting everything immediately?

A suite contains a set of Workload objects. The problem we had was this meant we needed to get all the resources (downloads and all) in the suite to just have a functioning suite object. This was of course incredibly inefficient when you only wanted to use one of these workloads.

The solution I implemented had the Workload objects constructed with a partial function which were essentially defining the function call needed to download all its components in our bucket. This partial function was only called when someone tried to accesses these resources via `get_local_path`. Ergo, you had a Suite of workloads, you could iterate through the workloads and inspect ids and versions as you wanted to, and only when you called `get_local_path` did it download what it needed for that workload.

I appreciate this solution is imperfect, but it saves us the bandwidth. 

While your solution in this PR solves pinging the database to death, does this not reintroduce the other bandwidth problem we were facing?

Reply:
This function also created the same partial as passes it to the constructor of the resource, so it will only download the required resource when `get_local_path` is called.

I am going to refactor the Workload and Suite resource so that we dont have to pass all the `obtain_resource` arguments like clients, resource_directory, etc to the constructors.